File: php/tests/Sample/Models/Transaction.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Sample\Models;

use SimpleOrm\Metadata\Attributes\Column;
use SimpleOrm\Metadata\Attributes\ForeignKey;
use SimpleOrm\Metadata\Attributes\Generated;
use SimpleOrm\Metadata\Attributes\Index;
use SimpleOrm\Metadata\Attributes\Key;
use SimpleOrm\Metadata\Attributes\ManyToOne;
use SimpleOrm\Metadata\Attributes\OneToMany;
use SimpleOrm\Metadata\Attributes\Table;
use SimpleOrm\Metadata\Attributes\Version;
use SimpleOrm\Query\SortOrder;
use SimpleOrm\Types\Decimal;

/** Table `transactions`: carries the version column — the optimistic-concurrency fixture (§7.16). */
#[Table('transactions')]
#[Index(['userId'])]
#[Index(['status', 'createdAtUtc', SortOrder::Desc], name: 'ix_transactions_status_created')]
final class Transaction extends BaseModel
{
    #[Key]
    #[Generated]
    #[Column]
    public int $id;

    #[Column]
    #[ForeignKey(User::class)]
    public int $userId;

    #[ManyToOne('userId')]
    public private(set) ?User $user = null;

    #[Column]
    public TransactionStatus $status = TransactionStatus::Pending;

    #[Column]
    public Decimal $amount;

    #[Version]
    #[Column]
    public int $version = 0;

    /** Added by migration V0003. */
    #[Column]
    public ?string $note = null;

    #[OneToMany(TransactionDetail::class, 'transactionId')]
    public private(set) array $details = [];
}

File: go/orm/sample/transaction.go
package sample

import "github.com/kosmassp/SimpleOrm/go/orm"

// Transaction is table transactions (STRICT). Key: id, database-generated;
// user_id references users. Carries the version column — the fixture entity
// for optimistic concurrency (§7.16).
type Transaction struct {
	BaseModel
	ID     int64 `orm:"column,key,generated"`
	UserID int64 `orm:"column"`
	// User is populated only by the library (Level 2 loading); it can never disagree with UserID.
	User    *User             `orm:"many_to_one=UserID"`
	Status  TransactionStatus `orm:"column"`
	Amount  orm.Decimal       `orm:"column"`
	Version int64             `orm:"column,version"`
	// Note was added by migration V0003.
	Note *string `orm:"column"`
	// Details is populated only by the library (Level 2 loading); never a column, never written.
	Details []*TransactionDetail `orm:"one_to_many=TransactionID"`
}

func (Transaction) Entity() orm.EntityDef {
	return orm.EntityDef{
		Source:      orm.Table("transactions"),
		ForeignKeys: []orm.ForeignKeyDef{orm.ForeignKey[User]("UserID")},
		Indexes: []orm.IndexDef{
			orm.Index("UserID"),
			orm.Index("Status", "CreatedAtUtc", orm.Desc).Named("ix_transactions_status_created"),
		},
	}
}
